framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java

```diff
@@ -26,17 +30,52 @@
 
     @Override
     public AnnotatedTypeMirror visitVariable(VariableTree node, AnnotatedTypeFactory f) {
+        Element elt = TreeUtils.elementFromDeclaration(node);
+
         // Create the ATM and add non-primary annotations
-        // (node.getType() does not include primary annotations, those are in
-        // node.getModifier()
+        // (node.getType() does not include the annotation before the type, so those
+        // are added to the type below).
         AnnotatedTypeMirror result = TypeFromTree.fromTypeTree(f, node.getType());
 
-        // Add primary annotations
+        // Handle any annotations in node.getModifiers().
+        List<AnnotationMirror> modifierAnnos;
         List<? extends AnnotationTree> annoTrees = node.getModifiers().getAnnotations();
         if (annoTrees != null && !annoTrees.isEmpty()) {
-            List<AnnotationMirror> annos = TreeUtils.annotationsFromTypeAnnotationTrees(annoTrees);
+            modifierAnnos = TreeUtils.annotationsFromTypeAnnotationTrees(annoTrees);
+        } else {
+            modifierAnnos = new ArrayList<>();
+        }
+
+        if (result.getKind() == TypeKind.DECLARED
+                &&
+                // Annotations on enum constants are not in the TypeMirror and always apply to the
+                // innermost type, so handle them in the else block.
+                elt.getKind() != ElementKind.ENUM_CONSTANT) {
+
+            // Decode the annotations from the type mirror because the annotations are already in
+            // the correct place for enclosing types. For example, @Tainted Outer.Inner and @Tainted
+            // Inner. For both types @Tainted is stored in node.getModifiers(), but they apply to
+            // different types.
+            AnnotatedDeclaredType annotatedDeclaredType = (AnnotatedDeclaredType) result;
+            // The underlying type of result does not have all annotations, but the TypeMirror of
+            // node.getType() does.
+            DeclaredType declaredType = (DeclaredType) TreeUtils.typeOf(node.getType());
+            AnnotatedTypes.applyAnnotationsFromDeclaredType(annotatedDeclaredType, declaredType);
+
+            // Handle declaration annotations
+            for (AnnotationMirror anno : modifierAnnos) {
+                if (AnnotationUtils.isDeclarationAnnotation(anno)) {
+                    // This does not treat Checker Framework compatqual annotations differently,
+                    // because it's not clear whether the annotation should apply to the outermost
+                    // enclosing type or the innermost.
+                    result.addAnnotation(anno);
+                } // If anno is not a declaration annotation, it should have been applied in the
```

Comment:
I don't understand why this comment is attached to the curly brace rather than on a separate line.  Can you clarify why?